crates/bevy_app/src/app.rs

```diff
@@ -229,171 +106,183 @@ impl App {
+    /// **This method is not required to return.**
     ///
-    /// In simple and *headless* applications, one can expect that execution will
-    /// proceed, normally, after calling [`run()`](App::run()) but this is not the case for
-    /// windowed applications.
+    /// Headless apps can generally expect this method to return control to the caller when
+    /// it completes, but that is not the case for windowed apps. Windowed apps are typically
+    /// driven by an event loop and some platforms expect the program to terminate when the
+    /// event loop ends.
     ///
-    /// Windowed apps are typically driven by an *event loop* or *message loop* and
-    /// some window-manager APIs expect programs to terminate when their primary
-    /// window is closed and that event loop terminates – behavior of processes that
-    /// do not is often platform dependent or undocumented.
-    ///
-    /// By default, *Bevy* uses the `winit` crate for window creation. See
-    /// [`WinitSettings::return_from_run`](https://docs.rs/bevy/latest/bevy/winit/struct.WinitSettings.html#structfield.return_from_run)
-    /// for further discussion of this topic and for a mechanism to require that [`App::run()`]
-    /// *does* return – albeit one that carries its own caveats and disclaimers.
+    /// Bevy uses `winit` as its default window manager. See [`WinitSettings::return_from_run`]
+    /// for a mechanism (with many additional caveats) that will guarantee [`App::run()`] returns
+    /// to the caller.
     ///
     /// # Panics
     ///
-    /// Panics if called from `Plugin::build()`, because it would prevent other plugins to properly build.
+    /// Panics if not all plugins have been built.
+    ///
+    /// [`WinitSettings::return_from_run`]: https://docs.rs/bevy/latest/bevy/winit/struct.WinitSettings.html#structfield.return_from_run
     pub fn run(&mut self) {
         #[cfg(feature = "trace")]
         let _bevy_app_run_span = info_span!("bevy_app").entered();
-
-        let mut app = std::mem::replace(self, App::empty());
-        if app.building_plugin_depth > 0 {
-            panic!("App::run() was called from within Plugin::build(), which is not allowed.");
+        if self.is_building_plugins() {
+            panic!("App::run() was called while a plugin was building.");
         }
 
-        if app.plugins_state() == PluginsState::Ready {
+        if self.plugins_state() == PluginsState::Ready {
             // If we're already ready, we finish up now and advance one frame.
             // This prevents black frames during the launch transition on iOS.
-            app.finish();
-            app.cleanup();
-            app.update();
+            self.finish();
+            self.cleanup();
+            self.update();
         }
 
-        let runner = std::mem::replace(&mut app.runner, Box::new(run_once));
+        let mut app = std::mem::replace(self, App::empty());
+        let runner = app.runner.take().expect("an app runner");
```

Comment:
This will panic if someone calls `run()` twice on the same app, right? Can we either improve this panic message or make this non-panicking?